src/helpers/coordinates.ts

```diff
@@ -1,164 +1,209 @@
+  string | { x: number; y: number; z: number } | [number, number, number] | Coordinates;
 
 export class Coordinates {
-  public x: number;
-  public y: number;
+  public readonly x: number;
+  public readonly y: number;
+  public readonly z: number;
+
+  /*
+  * A general introduction to how shape coordinates are handled:
+  *
+  * Shapes are placed on a grid. A shape is always rectangular, but does not have to be square.
+  * A shape consists of one or more squares.
+  * A square is always 1x1, and all four corners of a square have integer coordinates.
+  *
+  * The top left corner (lowest value x / y) of a shape or square is the anchor.
+  * As the anchor is a square corner, its coordinates are always integer values.
+  * We can infer that the center of a square always is [anchor.x + 0.5, anchor.y + 0.5].
+  *
+  * Edge points are indeterminate positions on the grid. Their coordinates values can be decimal in both X and Y.
+  *
+  * When rotating shapes, the shape anchor always retains the same global coordinates.
+  * Because all corners of squares must have integer coordinates, the smallest possible rotation is 90 degrees.
+  *
+  * To rotate a shape, we first rotate the shape around its anchor coordinate,
+  * and then shift it until the new anchor is at the same place as the old anchor.
+  *
+  * This shift will only ever be in the positive direction, on either or both X and Y axes.
+  * A small example of rotating a 3x2 shape 90 degrees:
+  *
+  * X marks the anchor position.
+  *
+  *  Start:
+  *              X
+  *               1 2 3
+  *               4 5 6
+  *
+  *  Rotate:
+  *              X
+  *           4 1
+  *           5 2
+  *           6 3
+  *
+  *  Shift:
+  *              X
+  *               4 1
+  *               5 2
+  *               6 3
+  */
 
   public constructor(param: CoordinatesParam) {
     if (typeof param === 'string') {
-      [this.x, this.y] = param.split(',').map(Number);
+      [this.x, this.y, this.z] = param.split(',').map(Number);
     }
 
     else if (Array.isArray(param)) {
-      [this.x, this.y] = param;
+      [this.x, this.y, this.z] = param;
     }
 
     else if (typeof param === 'object') {
       this.x = param.x;
       this.y = param.y;
+      this.z = param.z;
     }
 
     else {
       throw new Error(`${param} is not a valid argument`);
     }
 
-    if (typeof this.x !== 'number' || Number.isNaN(this.x)
-      || typeof this.y !== 'number' || Number.isNaN(this.y)) {
+    const rules = [
+      (v: number) => typeof v === 'number',
+      (v: number) => !Number.isNaN(v),
+    ];
+
+    if (!rules.every(rule => [this.x, this.y, this.z].every(v => rule(v)))) {
       throw new Error(`${param} could not be parsed as a coordinate`);
     }
   }
 
   private isException(): boolean {
-    return this.x < 0 || this.y < 0; // exclude negative coordinates which are used for exceptions
+    return this.z < 0; // negative Z values are used for static objects that should not move
```

Comment:
better explanation: negative z-values are used to categorize components that have 'special' coordinates that are not real grid coordinates, like system IO. These coordinates should not be shifted or rotated.